File: Web/src/features/harmonica/harmonicaAction.ts
import type { AreaSize } from './touch/areaSize.js'
import type { Contact } from './touch/contact.js'

export type HarmonicaAction =
    | { readonly kind: 'startButtonTapped' }
    | { readonly kind: 'pageHidden' }
    | { readonly kind: 'stripTouched'; readonly contacts: readonly Contact[]; readonly across: AreaSize }
    | { readonly kind: 'keySliderMoved'; readonly position: number }
    | { readonly kind: 'styleChosen'; readonly named: string }
    | { readonly kind: 'notesPerFingerChosen'; readonly notes: number }
    | { readonly kind: 'tuneChosen'; readonly index: number }
    | { readonly kind: 'stopTuneButtonTapped' }
    | { readonly kind: 'shapingPadTouched'; readonly contacts: readonly Contact[]; readonly across: AreaSize }
    | { readonly kind: 'shapingPadPinched'; readonly magnification: number }

File: Apps/iOS/Features/Harmonica/HarmonicaAction.swift
import CoreGraphics

enum HarmonicaAction {
    case appBecameActive
    case appLeftTheForeground
    case screenAppeared
    case screenDisappeared
    case stripTouched([FingerTouch], across: CGSize)
    case keySliderMoved(toPosition: Double)
    case tuneChosen(at: Int)
    case stopTuneButtonTapped
    case shapingPadTouched([FingerTouch], across: CGSize)
    case shapingPadPinched(by: CGFloat, within: CGSize)
    case settingsButtonTapped
}
